R/utilities.R

```diff
@@ -361,60 +363,90 @@ get_ssh_session = function(host="gphost01.bcgsc.ca"){
 gene_to_region = function(gene_symbol,
                           ensembl_id,
                           genome_build = "grch37",
-                          return_as = "region"){
-
+                          return_as = "region",
+                          sort_regions = TRUE,
+                          na_for_genes_not_found = FALSE){
+  
+  stopifnot('`genome_build` parameter must be "grch37" or "grch38"' = genome_build %in% c("grch37", "grch38"))
+  stopifnot('`return_as` parameter must be "region", "bed" or "df"' = return_as %in% c("region", "bed", "df"))
+  
   #set mart based on selected genome projection
   if(genome_build == "grch37"){
     gene_coordinates = grch37_gene_coordinates
     chr_select = paste0(c(c(1:22),"X","Y"))
-  }else if(genome_build == "hg38"){
+  }else{
     gene_coordinates = hg38_gene_coordinates
     chr_select = paste0("chr", c(c(1:22),"X","Y"))
   }
-
+  
   #filter on gene_symbol/ensembl_id
   if(!missing(gene_symbol) && missing(ensembl_id)){
-    gene_coordinates = dplyr::filter(gene_coordinates, hugo_symbol %in% gene_symbol)
+    gene_coordinates = gene_coordinates[ match(gene_symbol, gene_coordinates$hugo_symbol), ]
+    gene_coordinates$hugo_symbol <- gene_symbol
   }
-
+  
   if(missing(gene_symbol) && !missing(ensembl_id)){
-    gene_coordinates = dplyr::filter(gene_coordinates, ensembl_gene_id %in% ensembl_id)
+    gene_coordinates = gene_coordinates[ match(ensembl_id, gene_coordinates$ensembl_gene_id), ]
+    gene_coordinates$ensembl_gene_id <- ensembl_id
   }
-
+  
   region = dplyr::select(gene_coordinates, chromosome, start, end, gene_name, hugo_symbol, ensembl_gene_id) %>%
-    as.data.frame() %>%
-    dplyr::arrange(chromosome, start) %>%
-    dplyr::filter(chromosome %in% chr_select)
+    as.data.frame()
+  
+  if(sort_regions){
+    if(genome_build == "grch37"){
+      chrm_num = format( region$chromosome, justify = "right", na.encode = FALSE )
```

Comment:
Can you explain what the rationale is for these format lines? I don't know what this solves

Reply:
I think this change is for the correct ordering of the chromosomes (1,2,3 instead of 1,11,12..). But this format seems to add a prepending space at the front:

```
r$> chr <- c(rep(1, 5), rep(11, 3), rep("X", 5), rep(2, "5"))

r$> format(chr, justify = "right", na.encode = FALSE )
 [1] " 1" " 1" " 1" " 1" " 1" "11" "11" "11" " X" " X" " X" " X" " X" " 2" " 2" " 2" " 2" " 2"
```

I would  think we would avoid at all costs adding new dependencies because of how many there are already but another way to sort in the correct order is the 
```
r$> as.character(
        sort(
            factor(
                chr,
                levels = c((1:22), "X"),
                ordered = TRUE
            )
       )
    )
 [1] "1"  "1"  "1"  "1"  "1"  "2"  "2"  "2"  "2"  "2"  "11" "11" "11" "X"  "X"  "X"  "X"  "X" 
```

This will also work for the prefixed chromosomes

Reply:
@Kdreval, you are right, this is to sort the chromosomes. `format` function does add spaces to the strings, but this is not a problem here, because `format` output goes to a separate object, called `chrm_num`, that I use just for the sorting. The real chromosome values are unchanged in `region$chromosome`.

Reply:
`format` function is from the `base` packege.

Reply:
Does it still do a proper match later when the chromosomes are ordered? Just not sure how the comparison of ` 1` to `1` is made to ensure the correct order.

Reply:
Oh I see. I had missed this line earlier. 

```region = dplyr::arrange(region, chrm_num, start)```

@Kdreval shouldn't we be only doing this chr prefix removal for the grch37 case (i.e. in the else/hg38 context it should be left prefixed)?

Reply:
Right, this only does the `chr` handling for the else part of `if(genome_build == "grch37")` check, so I think this alright - but the order handling is done for both projections because when the `X` is present they are all treated as characters anyway and will result in that non-sensible 1,11,12..2... order

Reply:
I'm wondering out loud if this code and other code normalizing coordinates to builds could eventually shift into Laura's new process_coordinates helper function (still in a PR)

Reply:
I'm not using `format` function anymore, but @Kdreval suggestion to use chromosome numbers as ordered factors:
```
chrm_num = factor(chrm_num, levels = c(1:22, "X", "Y"), ordered = TRUE)
```